feature: DataGrid Columns allow Sort and Filter when bound to Array<o…

EdJohnstonG9 opened the pull request:
The objective is to use native display, filter and sort in a DataGrid<DataRow> 
This does not work atm as the DataRow.ItemArray[] returns object which cannot be filtered, but if DataColumn Type property is set then the propertyValueGetter is not set. The changes:

public static Type GetPropertyType(Type type, string property)// returns type of underlying array type when type==array<type>[]
GetNestedPropertyExpression(Expression expression, string property, Type type = null)// adds a Converter when property is array<T>[] and T != type
protected override void OnInitialized() // changes the logic here so that both Type can be set and propertyValueGetter is set under  these circumstances

The last is the change most needing of review. I have altered existing logic without fully understanding what it is handling. Maybe the logic should be added to handle the specific case where the underlying data object type != Type

enchev commented:
Please add the example to the ExmpleService - at the moment it's unreachable. When manually navigating to the example I've received the following error:
<img width="1919" height="1079" alt="image" src="https://github.com/user-attachments/assets/1d90c1be-8367-484c-8653-f7a9f7f59ef5" />

EdJohnstonG9 commented:
I have added the demo pages into ExampleService, name of "DataTable Source"
Apologies for the AmbiguousMatch; very embarrassing I left that in there on transferring changes to the fork.

New commit available

enchev commented:
All existing DataGrid features should work.

EdJohnstonG9 commented:
A quick update
I think I have fixed all the issues found so far and also simplified the solution (fewer changes)
However I do want to do more testing of scenarios like FilterProperty != Property
I am building some bUnit tests to cover these too, which is taking longer than expected...

EdJohnstonG9 commented:
I have worked my way through all the issues I can find:
- Display of all tested data, including Enums
- Handling nullable values including enum?
- Filtering and sorting in all modes
I have added some test which for separation, I have put in a separate folder, driven from various data sources
I found a few UI issues when mixed mode filtering, for example:
Having a CheckBoxList filter in mixed grid shows a Input box which 
Throws an error if entered
Shows the name of data object after selecting items from the list

I will see if I can find the issue and suggest a fix.

enchev commented:
I received the follwoing error while tried the demo without changing anything:
<img width="1919" height="1079" alt="image" src="https://github.com/user-attachments/assets/871d9ec0-bbf0-4776-a896-29b90de80ca2" />

<img width="1919" height="1079" alt="image" src="https://github.com/user-attachments/assets/f38f87db-73d0-498a-8cd0-fc4543fc8754" />

EdJohnstonG9 commented:
Yes that is the case I mentioned above
When a column is configured for checkbox list and others or the grid for Simple
Then the input box appears
And using it gives this error

same issue on the current live version
Thinking the best solution is to prevent render of input box in this case. Or instead render dropdown select box

enchev commented on Radzen.Blazor/RadzenDataGrid.razor:
There is already fix for this in the master branch:
https://github.com/radzenhq/radzen-blazor/commit/206daba741121721a5ea1276f7ee7157fa9945fb

EdJohnstonG9 commented:
I have found one issue which is not the cause and a theory that I cannot find a way to prove easily

The issue is changing from one Filter Mode to the other, does not update the FilterPropertyType to/from IEnumerable:

- When chagning to FilterMode.CheckBoxList from any other, the column needs to know the FilterPropertyType is an IEnumberable
- However this is only set during OnInitialize which is not called under those circumstances
- The reverse is true also, changing from CheckBoxList to any other
- This could result in some issues with the current live code but I think is accidentally improved with the Fork I am working on

Testing with the initial grid FilterMode = CheckBoxList, the issue persists but with the Expression created by public static IQueryable<T> Where<T>(.. being equivalent to simple grids

That leaves only this code in QueryableExtension.cs: 

```
public static IQueryable<T> Where<T>(this IQueryable<T> source,...
            var lambda = Expression.Lambda<Func<T, bool>>(combinedExpression, parameter);

            return source.Where(lambda);
```

It looks like the Where(lambda) is operating recursively, finding a parent or child Employee through Employee1/Employees1 which in turn is finding parent/child Employee etc, rather than operating only on the primary Data<Employee> enumerable/queryable.

However I cannot yet work out for sure if this is happening or how to stop it if it is.

enchev commented:
In my opinion these issues are unrelated to the DataTable binding and we can fix them in other iterations. I've not found anything specific that might cause a problem with this pull request and we ca merge it if you think you are ready.

EdJohnstonG9 commented:
Hi @enchev anything holding this back from merging?
Let me know if you need anything more from me.

EdJohnstonG9 commented:
I was trying to work out how you did this so easily, and this just solves so many issues right off the bat
`
public static string GetDynamicPropertyExpression(string name, Type type)
`

However there is still an issue rendering Enums which currently are displaying as their Int value not the representation.

GetDynamicPropertyExpression deliberately returns (Enum)it[{property}] for all enums, rather than the underlying type- not sure why. 
Even if changed, the function GetNestedPropertyExpression cannot add the relevant converter of int->{myEnumType}

I resolved this in a very clumsy way but sure you must know a better way.
